traits/observers/_trait_event_notifier.py

```diff
@@ -0,0 +1,207 @@
+        count is bumped. Hence this method is not idempotent.
+        N number of calls to this ``add_to`` must be matched by N calls to the
+        ``remove_from`` method in order to completely remove a notifier from
+        an observable.
+
+        Parameters
+        ----------
+        observable : IObservable
+            An object for adding this notifier to.
+
+        Raises
+        ------
+        RuntimeError
+            If the internal reference count is not zero and an equivalent
+            notifier is not found in the observable.
+        """
+        notifiers = observable._notifiers(True)
+        for other in notifiers:
+            if self.equals(other):
+                other._ref_count += 1
+                break
+        else:
+            # It is not a current use case to share a notifier with multiple
+            # observables. Using a single reference count will tie the lifetime
+            # of the notifier to multiple objects.
+            if self._ref_count != 0:
+                raise RuntimeError(
+                    "Sharing notifiers across observables is unexpected."
+                )
+            notifiers.append(self)
+            self._ref_count += 1
+
+    def remove_from(self, observable):
+        """ Remove this notifier from an observable object.
+
+        If an equivalent notifier exists, the existing notifier's reference
+        count is decremented and the notifier is only removed if
+        the count is reduced to zero.
+
+        Parameters
+        ----------
+        observable : IObservable
+            An object for removing this notifier from.
+
+        Raises
+        ------
+        RuntimeError
+            If the reference count becomes negative unexpectedly.
+        ValueError
+            If the notifier is not found.
+        """
+        notifiers = observable._notifiers(True)
+        for other in notifiers[:]:
+            if self.equals(other):
+                if other._ref_count == 1:
+                    notifiers.remove(other)
+                other._ref_count -= 1
+                if other._ref_count < 0:
+                    raise RuntimeError(
+                        "Reference count is negative. "
+                        "Race condition?"
+                    )
+                break
+        else:
+            # We may have to relax this later when dealing with
+            # "old" default value that don't have any notifiers.
+            raise ValueError("Notifier not found.")
+
+    def equals(self, other):
+        """ Return true if the other notifier is equivalent to this one.
+
+        Parameters
+        ----------
+        other : any
+
+        Returns
+        -------
+        boolean
+        """
+        if other is self:
+            return True
+        if type(other) is not type(self):
+            return False
+        return (
+            self.handler() is other.handler()
+            and self.target() is other.target()
```

Comment:
Another discussion point: I think currently if one calls `on_trait_change` with the same arguments except for `dispatch`, the first one wins:
```
foo.on_trait_change(handler, "name", dispatch="same")
foo.on_trait_change(handler, "name", dispatch="ui")
```
The second call is ignored because "dispatch" is not considered for differentiating notifiers. Here we could differentiate them, and I am tempted to think that we should. In other words, calling the following will result in two notifiers, one to be dispatched on the same thread, and the second one to be dispatched on the GUI event loop:
```
obj.observe(handler, "name", dispatch="same")
obj.observe(handler, "name", dispatch="ui")
```

Reply:
I am in favour of this.

Reply:
:+1: I will add this one line for comparing the dispatch callable (and its test).